### ISSUE TITLE
Allow assignment to the local variable created by *ngFor.

### ISSUE THREAD

**Issue by uber5001** [angular/angular#17144]
**I'm submitting a feature request**

**Current behavior**
A common pattern throughout my application is binding to `ngModel` within an *ngFor:
<pre>
&lt;input *ngFor="let foo of listOfFoos; let i = index" [(ngModel)]="listOfFoos[i]">
</pre>
You'll notice I don't even use `foo` here.

**Proposed behavior**
Allow assignment to the local variable `foo` created by *ngFor.
<pre>
&lt;input *ngFor="<b>bind foo</b> of listOfFoos" [(ngModel)]="<b>foo</b>">
</pre>
Keeping the `let` syntax would be confusing for this feature, as `let` implies a local variable that has no side effects upon being updated. A variable with `bind` would have the side effect of assigning to it's original source, more like an alias than a variable. If this is implemented, a `bind` variable would likely require a `trackBy` to know what index to assign its new values to.

**Minimal reproduction of the problem with instructions**
[Here's a simple example on plunker](https://plnkr.co/edit/jgYbUzTy8q3l1fpCJIE9?p=preview). The desired behavior is commented out. Note that I included `trackBy` to get this example to work, and I don't expect that to go away with the new proposed behavior.

**What is the motivation / use case for changing the behavior?**
Being able to bind to the array element's *ngFor variable (`foo`, in this case) seems like it should "just work", like in #8602 and #8086.

It also looks like this was possible in angularJS. Adding this feature would make angularJS -> Angular migration that much easier.

**Please tell us about your environment:**

* **Angular version:** 4.1.3

* **Browser:** all

 
* **Language:** TypeScript 2.1

* **Node (for AoT issues):** `node --version` =   `v6.10.2`

**Comment by trotyl** [comment]
Possibly duplicate of #8602.

**Comment by pkozlowski-opensource** [comment]
In your case what you _really_ want to do is to bind to the _element in a list_ and as such `listOfFoos[i]` _must_ show up somewhere. Once again, the intention is to assign to an element of an existing list and _not_ to a copy (reference) to one particular element. 

> Allow assignment to the local variable foo created by *ngFor.

IMO this would be counter-intuitive as the equivalent in JS would be to do sth along those lines:

```js
const items = [1, 2, 3];
for (let item of items) { 
    item = item + 1; 
}
```

and _expect_ that `items` will have `[2, 3, 4]`. This obviously won't work.

What you _must_ do is:

```js
const items = [1, 2, 3];
for (let i=0; i < items.length; i++) { 
    items[i] = items[i] + 1;
}
```

We don't want `ngFor` / `ngModel` to behave differently as compared to pure JS as this could become confusing pretty quickly.

> You'll notice I don't even use foo here.

If this local variable bothers you there is the de-sugared syntax that you could use (which is closer to the for loop with indexes):

```html
<ng-template ngFor [ngForOf]="listOfFoos" let-i="index">
    <input [(ngModel)]="listOfFoos[i]">
</ng-template>
``` 

which is closer to the JS loop (not saying that it is easier to write but probably makes the intention clearer).

tl;dr; I understand what you are asking for but we will pass on the suggestion for now as it would introduce syntax and constructs that would make us (even) more different from the standard JS behaviour.
